File: OldCcode/jan20files/heatapp.c
/* ---------------------------------------------------------------------------------- */
/*                                                                                    */
/*                                    HEAT.C an Approximation                         */
/*                                                                                    */
/* ---------------------------------------------------------------------------------- */

/* ---------------------------------------------------------------------------------- */
/* This function is the new fifth part of the objective function that deals with      */
/* the temperature of components.  If the component's temperature exceeds the         */
/* critical temperature of that component then a penalty will result.                 */
/* ---------------------------------------------------------------------------------- */
void eval_part_5(design)
struct Design *design;
{
  int i;
  struct Component *comp;
  double calc_thermal_penalty(), calc_thermal_matrix();
  double tempave;

/*  printf("Calculating Thermal Matrix\n");*/
  tempave = calc_thermal_matrix(design);
/*  printf("%.2f\n", tempave);*/
  design->new_obj_values[3] = 0.0;
  comp = design->first_comp;
  for (i = 1; i <= COMP_NUM; ++i) {
    design->new_obj_values[3] = calc_thermal_penalty(tempave, comp->tempcrit);
    if (i < COMP_NUM)
      comp = comp->next_comp;
  }
}

/* ---------------------------------------------------------------------------------- */
/* This function returns the value of the penalty function for a temperature in       */
/* excess of the critical temperature.                                                */
/* ---------------------------------------------------------------------------------- */
double calc_thermal_penalty(temp, tempcrit)
double temp, tempcrit;
{
  double value;

  value = 0.0;
  value = (temp*temp)/(tempcrit*tempcrit*COMP_NUM);
#ifdef MINIMIZE
  if (temp > tempcrit)
    value += (temp - tempcrit)*(temp - tempcrit)/(COMP_NUM);
#endif
  return(value);
}

/* ---------------------------------------------------------------------------------- */
/* This function finds the temperature at the center of each component and places     */
/* that value in comp->temp.  Because of the nature of this function it needs to      */
/* re-calculated for each iteration, instead of just being updated.                   */
/* ---------------------------------------------------------------------------------- */
double calc_thermal_matrix(design)
struct Design *design;
{
  struct Component *comp;
  double Tave, Rtot, Qtot = 0.0, Kave = 0.0, Have = 0.0;
  double box_x_dim, box_y_dim, box_z_dim, box_area, box_volume;
  int i;


  box_x_dim = design->box_max[0] - design->box_min[0];
  box_y_dim = design->box_max[1] - design->box_min[1];
  box_z_dim = design->box_max[2] - design->box_min[2];

  box_volume = box_x_dim * box_y_dim * box_z_dim;
  box_area = 2*(box_x_dim*box_y_dim + box_x_dim*box_z_dim + box_y_dim*box_z_dim);

  comp = design->first_comp;
  for (i = 1; i <= COMP_NUM; ++i) {
    Qtot += comp->q;
    Kave += (comp->k)/COMP_NUM;
    comp = comp->next_comp;
  }
  Kave = Kave*(design->volume/box_volume) + (design->kb)*(1 - design->volume/box_volume);

  Have = ((design->hx) + (design->hy) + (design->hz))/DIMENSION;

  Rtot = (box_area/(Kave*box_volume)) + 1/(Have*box_area);
  Tave = (design->tamb) + (Qtot*Rtot);

  return(Tave);
}

/* ---------------------------------------------------------------------------------- */
/* This function reverts to  the previous node temperatures if the new move was       */
/* rejected.                                                                          */
/* ---------------------------------------------------------------------------------- */
void revert_tfield(design)
struct Design *design;
{
  int k;

  for (k = 0; k < NODE_NUM; ++k)
    design->tfield[k].temp = design->tfield[k].old_temp;
}

/* ---------------------------------------------------------------------------------- */
/* This function backs up the current temperatures into old_temp if the step was      */
/* accepted.                                                                          */
/* ---------------------------------------------------------------------------------- */
void back_up_tfield(design)
struct Design *design;
{
  int k;

  for (k = 0; k < NODE_NUM; ++k)
    design->tfield[k].old_temp = design->tfield[k].temp;
}

File: 3D_LayoutOpt/Functions/HeatAPP.cs
namespace _3D_LayoutOpt
{
    class HeatApp
    {

        /* ---------------------------------------------------------------------------------- */
        /*                                                                                    */
        /*                        HEATAPP.C -- Approximation Method                           */
        /*                                                                                    */
        /* ---------------------------------------------------------------------------------- */

        /* ---------------------------------------------------------------------------------- */
        /* This function corrects the approximation method by comparing it to the LU method.  */
        /* The value for design.hcf (heat correction factor) will be used by the app_method. */
        /* ---------------------------------------------------------------------------------- */
        public static void correct_APP_by_LU(Design design)
        {
            Component comp;
            double tempapp, tempMm;
            tempMm = 0.0;
            design.Hcf = 1.0;
            design.Gauss = 0;
            ThermalAnalysisAPP(design);
            tempapp = design.Components[0].Temp;

            HeatMm.ThermalAnalysisMM(design);

            for (var i = 0; i < design.CompCount; i++)
            {
                comp = design.Components[i];
                if (tempMm < comp.Temp)
                    tempMm = comp.Temp;
            }
            design.Hcf = tempMm/tempapp;
        }

        /* ---------------------------------------------------------------------------------- */
        /* This function finds the temperature at the center of each component and places     */
        /* that value in comp.temp.  Because of the nature of this function it needs to      */
        /* re-calculated for each iteration, instead of just being updated.                   */
        /* ---------------------------------------------------------------------------------- */
        public static void ThermalAnalysisAPP(Design design)
        {
            Component comp;
            double tave, rtot, qtot = 0.0, kave = 0.0, have = 0.0;
            double boxXDim, boxYDim, boxZDim, boxArea, boxVolume;

            boxXDim = design.BoxMax[0] - design.BoxMin[0];
            boxYDim = design.BoxMax[1] - design.BoxMin[1];
            boxZDim = design.BoxMax[2] - design.BoxMin[2];

            boxVolume = boxXDim* boxYDim * boxZDim;
            boxArea = 2*(boxXDim* boxYDim + boxXDim* boxZDim + boxYDim* boxZDim);

            for (var i = 0; i < design.CompCount; i++)
            {
                comp = design.Components[i];
                qtot += comp.Q;
                kave += (comp.K) / design.CompCount;
            }

            kave = kave*(design.Volume/boxVolume) + (design.Kb)*(1 - design.Volume/boxVolume);

            have = ((design.H[0]) + (design.H[1]) + (design.H[2]))/Constants.Dimension;

            /*  Rtot = (box_area/(Kave*box_volume)) + 1/(Have*box_area);*/
            rtot = ((boxXDim + boxYDim + boxZDim)/(kave* boxArea)) + 1/(have* boxArea);
            tave = (design.Tamb) + design.Hcf*(qtot* rtot);


            for (var i = 0; i < design.CompCount; i++)
            {
                comp = design.Components[i];                                                                            //SETTING Tave TO ALL COMPONENTS
                comp.Temp = tave;
            }
        }
    }
}
